src/main/kotlin/org/simplemc/simpleannounce/config/SimpleAnnounceConfig.kt

```diff
@@ -0,0 +1,135 @@
+package org.simplemc.simpleannounce.config
+
+import com.fasterxml.jackson.annotation.JsonAlias
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.annotation.JsonTypeInfo
+import com.fasterxml.jackson.annotation.JsonUnwrapped
+import org.bukkit.boss.BarColor
+import org.bukkit.boss.BarStyle
+import org.simplemc.simpleannounce.inTicks
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+data class SimpleAnnounceConfig(
+    val autoReload: Duration? = null,
+    val announcements: List<AnnouncementConfig<*>>,
+) {
+    companion object {
+        private fun Duration?.checkNullZeroOrPositive(name: String) {
+            check(this == null || this == Duration.Companion.ZERO || this.isPositive()) {
+                "When set, $name must be >= 0s"
+            }
+        }
+    }
+
+    @JsonIgnore
+    val autoReloadTicks = autoReload?.inTicks
+
+    init {
+        check(autoReload == null || autoReload == Duration.Companion.ZERO || autoReload.inWholeMinutes >= 1) {
+            "When set, Auto Reload Duration must be > 1 minute"
+        }
+    }
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.SIMPLE_NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
+    sealed class AnnouncementConfig<T> {
+        abstract val random: Boolean
+        abstract val delay: Duration
+        abstract val repeat: Duration?
+        abstract val includesPermissions: List<String>
+        abstract val excludesPermissions: List<String>
+        abstract val messages: List<T>
+
+        @JsonIgnore
+        val delayTicks = delay.inTicks.toInt()
+
+        @JsonIgnore
+        val repeatTicks = repeat?.inTicks?.toInt()
+
+        init {
+            delay.checkNullZeroOrPositive("delay")
+            repeat.checkNullZeroOrPositive("repeat")
+        }
+
+        data class Chat(
+            override val random: Boolean = false,
+            override val delay: Duration = Duration.Companion.ZERO,
+            override val repeat: Duration? = null,
+            override val includesPermissions: List<String> = emptyList(),
+            override val excludesPermissions: List<String> = emptyList(),
+            @field:JsonAlias("message") override val messages: List<String>,
+        ) : AnnouncementConfig<String>()
+
+        data class Boss(
+            override val random: Boolean = false,
+            override val delay: Duration = Duration.Companion.ZERO,
+            override val repeat: Duration? = null,
+            override val includesPermissions: List<String> = emptyList(),
+            override val excludesPermissions: List<String> = emptyList(),
+            @field:JsonAlias("message") override val messages: List<BossBarMessage>,
+            @field:JsonUnwrapped val barConfig: BarConfig = BarConfig(),
```

Comment:
I could have used some other kotlin config library like `hoplite` instead of `jackson` and avoided a small portion of the pain (with an old version of jackson being included with spigot), but would have still had to deal with the `kotlin-reflect` problems and would lose the ability to `unwrap` properties like I do here.

...this makes it so the announcement configuration for a boss-type announcement has the `barConfig` fields directly (eg `hold`) instead of nested inside a `barConfig` field/object. Same with them being on the `messages` themselves.

Reply:
May or may not be interesting for you, but [Configurate](https://github.com/SpongePowered/Configurate/blob/trunk/README.md) is a popular config library for modern plugins (Paper uses it)

Reply:
[hoplite](https://github.com/sksamuel/hoplite) is simpler and easier to use IMO, but both would lose some of the jackson niceties (although Configurate seems to use jackson for its JSON support). Configurate also [uses kotlin-reflect](https://github.com/SpongePowered/Configurate/blob/trunk/extra/kotlin/build.gradle#L63) for its kotlin data class support, so many of the same things would need to be addressed anyway.